File: core-java/src/main/java/chessgame/domain/model/Queen.java
package chessgame.domain.model;

import java.util.LinkedHashSet;
import java.util.Set;

/**
 * Queen of Chess game.
 *
 * @author Prestes, E. M.
 * @see chessgame.domain.Piece
 * @since September 2016
 */
public class Queen extends AbstractPiece {

    /**
     * Queen's constructor.
     *
     * @param board Board informed.
     * @param color Color informed.
     * @see BoardPosition
     * @see PieceColor
     */
    public Queen(Board board, PieceColor color) {
        super(board, color);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Set<BoardPosition> getAvailablePositions() {
        final BoardPosition myPosition = getPosition();
        final Set<BoardPosition> availablePositions = new LinkedHashSet<>();

        availablePositions.addAll(collectColumns(myPosition));
        availablePositions.addAll(collectRows(myPosition));
        availablePositions.addAll(collectDiagonals(myPosition));

        return availablePositions;
    }
}

File: core-kotlin/src/main/kotlin/chessgame/domain/model/Queen.kt
package chessgame.domain.model

class Queen(
        override val board: Board,
        override val color: PieceColor) : AbstractPiece(board, color) {

    override fun getAvailablePositions(): Set<BoardPosition> {
        val availablePositions = mutableSetOf<BoardPosition>()

        collectColumns(position, availablePositions)
        collectRows(position, availablePositions)
        collectDiagonals(position, availablePositions)

        return availablePositions
    }
}